text/0082-cep-78-trading-extension.md

```diff
@@ -0,0 +1,115 @@
+# CEP-82 - Trading Extension for CEP-78
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#82](https://github.com/casperlabs/ceps/pull/82)
+
+Trading capabilities for CEP-78 based NFTs and associated conventions.
+
+## Motivation
+
+[motivation]: #motivation
+
+The goal of this CEP is to explore the current capabilities of CEP-78 to facilitate NFT trading use cases, and to outline flows and interfaces that could be standardized for greater interoperability of different marketplace implementations and CEP-78 NFTs.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+We introduce a new contract type called a `Marketplace`, the main purpose of which is to facilitate the exchange of NFTs between interested parties. There are two primary cases to consider:
+* Exchanging NFTs for other NFTs,
+* Exchanging NFTs for fungible tokens.
+
+### Custody
+
+The `Marketplace` contract can control the exchanged NFTs by either:
+* Having ownership of the NFT transferred to it directly,
+* Becoming an `operator` of the NFT via approval.
+
+Which one is more appropriate will depend on the marketplace and its model, and can be controlled through a _modality_ on the contract. The presence of this modality is also important for informing the consumers of the `Marketplace` contract on which mechanism to use.
+
+* `CustodyApproval` - custody via approval
+* `CustodyTransfer` - custody via transfer
+
+Taking ownership of the NFT is needed in cases where it is necessary to prevent the owner of the NFT from unexpectedly transferring the NFT to a third party, such as during an auction. Otherwise, approvals are more appropriate, as they do not change the actual owner of the NFT. 
+
+_(wip) remark: this requirement might conflict with the potential implementation of forced royalty collection using an intermediary custodial/royalties contract_
+
+### Duration
+
+Another dimension that can differentiate `Marketplace` contracts is _duration_:
+* A `Marketplace` can be _instant_, meaning that an exchange of items can happen as soon as two parties agree, such as in a swap. A consequence of this is that only the _seller_ needs to hand over custody of the item being sold, whereas the _buyer_ operation can be atomic. `CustodyApproval` will suffice for this `Marketplace`
```

Comment:
If I'm reading this correctly, then with only `CustodyApproval` I think there is potential for the order book to become out of sync if the NFT owner directly transfers the item while it is listed, which has several ramifications depending on the order book implementation.

Reply:
This is true. To be fair, I'm struggling to find a use case where using the `approve` mechanism would make sense.

Though in a very simplistic implementation, the offending item could simply be removed upon trying to purchase it.

Reply:
Yes but doesn't lead to the best market UX - user tries to buy, the transaction fails, they get charged a fee & find out the item isn't actually for sale 😓  

Also for 3rd party dapps & services they would not have access to a correct orderbook while items are incorrectly listed. Assuming there will be an orderbook index to access.